pep-0570.rst

```diff
@@ -385,6 +385,202 @@ as a rule is not LL(1). This is the reason the rule needs to be included in
+Let's say that you want your function to receive an indeterminate number of parameters,
+so the function will perform the operation with all of them, instead of calling the
+function multiple times with each pair. The signature will be now ::
+
+    def hilbert_multiply(*tensors):
+        ....
+
+Notice that if you used previously positional-only parameters, this is a backwards
+compatible change as the language is making sure that users were not using the name
+of the parameters when calling the function. So now, the old way of calling and the
+new way of calling are still possible ::
+
+    >>> hilbert_multiply(elem1, elem2) # Old way of calling the function
+
+    >>> hilbert_multiply(elem1, elem2, elem3, elem 4) # New way of calling the function
+
+If the function were not declared previously using positional-only parameters, then
+users will receive an error when calling the function as they used to do ::
+
+    >>> hilbert_multiply(x1=elem1, x2=elem2)
+    Traceback (most recent call last):
+      File "<stdin>", line 1, in <module>
+      TypeError: hilbert_multiply() got an unexpected keyword argument 'x1'
+
+
+There are more scenarios when positional-only parameters are beneficial. One of such
+scenarios is where a function can take some parameter and *any* keyword. For instance,
+let's imagine that we need a function that is able to format some template object. You
+can start to write ::
+
+    def format_template(template, **kwargs):
+        if isinstance(template, FastTemplate):
+            return template.format_fast(**kwargs)
+        elif isinstance(template, SlowTemplate):
+            return template.format_slow(**kwargs)
+        else:
+            return template.generic_format(**kwargs)
+
+Now you can call you function as ::
+
+    >>> format_template(slow_template, name="Hilbert", manifold="hausdorff")
+
+The problem with this approach is that you will never be able to pass the keyword
+"template" as the function will bind that value to the first argument ::
+
+    format_template(slow_template, name="Hilbert", template="fractal")
+    Traceback (most recent call last):
+      File "<stdin>", line 1, in <module>
+      TypeError: format_template() got multiple values for argument 'template'
+
+If you declare your function using positional-only parameters, this problem will go
+away as the first argument (template) is just a mere placeholder for the "template"
+object ::
+
+    def format_template(template, /, **kwargs):
+        if isinstance(template, FastTemplate):
+            return template.format_fast(**kwargs)
+        elif isinstance(template, SlowTemplate):
+            return template.format_slow(**kwargs)
+        else:
+            return template.generic_format(**kwargs)
+
+Now users can call your function with *all* possible keyword parameters ::
+
+    >>> format_template(slow_template, name="Hilbert", template="fractal")
+    "After some slow computations, *Hilbert* has finished the *fractal* template"
+
```

Comment:
I would remove the alternative section. If we have the syntax users win nothing by knowing how to hack it around. I think it is just noise.

Reply:
I think it helps to realize what are the advantages. Users are already used to `*args` and the hacks with that, so this will help to show how those constructs can be substituted with positional-only.

Reply:
This is a draft tutorial aimed at new users - it needs to teach them what the syntax does, it doesn't need to teach them all the API design trade-offs that might lead them to choosing whether or not to use it. Instead, it can just give a couple of specific examples of standard library APIs that work that way, like `dict.update`.

Reply:
It would be good to split out the alternative/advanced usage into its own section. It's useful to document, similar to when we rewrote the asyncio docs, what concepts most users would need from those that advanced users or library authors might need. This seems like a similar case here. The downside of not documenting advanced alternatives (even if we discourage them) is that it leaves it open for interpretation what the intent is and when to use (which the community will likely fill in with multiple different ways).